tests/devices/test_default_qubit_jax.py

```diff
@@ -0,0 +1,462 @@
+        def circuit(x):
+            qml.RX(x[1], wires=0)
+            qml.Rot(x[0], x[1], x[2], wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        res = circuit(p)
+
+        expected = jnp.cos(y) ** 2 - jnp.sin(x) * jnp.sin(y) ** 2
+        assert jnp.allclose(res, expected, atol=tol, rtol=0)
+
+        grad_fn = jacobian_transform(circuit, 0)
+        res = grad_fn(p)
+
+        expected = jnp.array(
+            [-jnp.cos(x) * jnp.sin(y) ** 2, -2 * (jnp.sin(x) + 1) * jnp.sin(y) * jnp.cos(y), 0]
+        )
+        assert jnp.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_state_differentiability(self, tol):
+        """Test that the device state can be differentiated"""
+        dev = qml.device("default.qubit.jax", wires=1)
+
+        @qml.qnode(dev, diff_method="backprop", interface="jax")
+        def circuit(a):
+            qml.RY(a, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        a = jnp.array(0.54)
+
+        def cost(a):
+            """A function of the device quantum state, as a function
+            of ijnput QNode parameters."""
+            circuit(a)
+            res = jnp.abs(dev.state) ** 2
+            return res[1] - res[0]
+
+        grad = jax.grad(cost)(a)
+        expected = jnp.sin(a)
+        assert jnp.allclose(grad, expected, atol=tol, rtol=0)
+
+    def test_prob_differentiability(self, tol):
+        """Test that the device probability can be differentiated"""
+        dev = qml.device("default.qubit.jax", wires=2)
+
+        @qml.qnode(dev, diff_method="backprop", interface="jax")
+        def circuit(a, b):
+            qml.RX(a, wires=0)
+            qml.RY(b, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return qml.probs(wires=[1])
+
+        a = jnp.array(0.54)
+        b = jnp.array(0.12)
+
+        def cost(a, b):
+            prob_wire_1 = circuit(a, b).squeeze()
+            return prob_wire_1[1] - prob_wire_1[0]
+
+        res = cost(a, b)
+        expected = -jnp.cos(a) * jnp.cos(b)
+        assert jnp.allclose(res, expected, atol=tol, rtol=0)
+
+        grad = jax.jit(jax.grad(cost, argnums=(0, 1)))(a, b)
+        expected = [jnp.sin(a) * jnp.cos(b), jnp.cos(a) * jnp.sin(b)]
+        assert jnp.allclose(grad, expected, atol=tol, rtol=0)
+
+    def test_backprop_gradient(self, tol):
+        """Tests that the gradient of the qnode is correct"""
+        dev = qml.device("default.qubit.jax", wires=2)
+
+        @qml.qnode(dev, diff_method="backprop", interface="jax")
+        def circuit(a, b):
+            qml.RX(a, wires=0)
+            qml.CRX(b, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+        a = jnp.array(-0.234)
+        b = jnp.array(0.654)
+
+        res = circuit(a, b)
+        expected_cost = 0.5 * (jnp.cos(a) * jnp.cos(b) + jnp.cos(a) - jnp.cos(b) + 1)
+        assert jnp.allclose(res, expected_cost, atol=tol, rtol=0)
+        # TODO(chase): Get grad to work without needing the wrapping.

```

Comment:
Still a TODO?

Reply:
How come the wrapping is needed?

Reply:
Wrapping is needed because JAX can only apply `jax.grad` to methods that return strict scalar values. Non-tape mode qnodes return tensors of shape (1,), and JAX complains. This wrapping fixing it.

Reply:
Removed TODO since this is fixed by tape mode.